### ISSUE TITLE
[Question] This works but the console log suggests it shouldn't?

### ISSUE THREAD

**Issue by JobLeonard** [joblib/joblib#507]
Well, this must be the nicest bug ever... something working when it says it shouldn't!

The set-up: SPA data browser, serving lists of individual rows of a laaaarge dataset as JSON through a flask server. The dataset is stored as an HDF5 file, so we're using numpy and h5py to fetch these rows on the fly.

Here's what we used to have:

```python
# Get one or more rows of data (i.e. all the expression values for a single gene)
@app.route('/loom/<string:project>/<string:filename>/row/<intdict:rows>')
@cache(expires=None)
def send_row(project, filename, rows):
	(u,p) = get_auth(request)
	ds = app.cache.connect_dataset_locally(project, filename, u, p)
	if ds == None:
		return "", 404
	else:
		# return a list of {idx, data} objects.
		# This is to guarantee we match up row-numbers client-side
		retRows = [{ 'idx': row, 'data': ds[row, :].tolist()} for row in rows]
		return flask.Response(ujson.dumps(retRows), mimetype="application/json")
```

Here's a stress-test with a relatively large file request:
```
127.0.0.1 - - [2017-03-03 14:48:37] "GET <url to fetch 300+ rows adding up to 1MB of data> HTTP/1.1" 200 1000983 14.084581
```
Less than a megabyte of data used to take 14.08 seconds!

Here's what I got after half an hour of messing around with joblib:

```python

# Helper functions to quickly fetch and convert a whole bunch of rows
def getRow(ds, row):
	return ds[row, :]

def npToList(row, npArray):
	return npArray.tolist()

def makeRow(rows, listRows, i):
	return {'idx': rows[i], 'data': listRows[i]}

def makeRows(ds, rows):
	with Parallel(n_jobs=4) as parallel:
		npRows = parallel(delayed(getRow)(ds, row) for row in rows)
		listRows = parallel(delayed(npToList)(rows, npArray) for npArray in npRows)
		return parallel(delayed(makeRow)(rows, listRows, i) for i in range(len(rows)))

# Get one or more rows of data (i.e. all the expression values for a single gene)
@app.route('/loom/<string:project>/<string:filename>/row/<intdict:rows>')
@cache(expires=None)
def send_row(project, filename, rows):
	(u,p) = get_auth(request)
	ds = app.cache.connect_dataset_locally(project, filename, u, p)
	if ds == None:
		return "", 404
	else:
		# return a list of {idx, data} objects.
		# This is to guarantee we match up row-numbers client-side
		return flask.Response(ujson.dumps(makeRows(ds, rows)), mimetype="application/json")
```

And here's the output:
```
/home/job/Dev/anaconda3/lib/python3.5/site-packages/loompy-0.22.26-py3.5.egg/loompy/loom_server.py:217: UserWarning: Multiprocessing backed parallel loops cannot be nested below threads, setting n_jobs=1
  with Parallel(n_jobs=4) as parallel:
127.0.0.1 - - [2017-03-03 14:49:34] "GET <url to fetch 300+ rows> HTTP/1.1" 200 119945 0.456284
```

BOOM! Half a second! Great, so what's the issue? Well, that message saying `setting n_jobs=1`. See, I figured that if it doesn't *really* do multi-processing, I might as well explicitly set the `n_jobs` back to 1. But if I do that, it's 14+ seconds again.

So my question is: why does this work regardless of this message? And why does manually setting `n_jobs=1` undo the over 30x improvement?

**Comment by JobLeonard** [comment]
Ah, celebrated too soon: it bails out after about five rows. Yeah, I can speed up fetching 300 rows that way too ;)

Oh well, at least the refactoring shaved off a second of time.

**Comment by lesteve** [comment]
Can you try with the latest released joblib (0.11) and see whether you still get the warnings. It could well be related to https://github.com/joblib/joblib/pull/414.

It is rather hard to help you because there is a lot of stuff involved in your snippet and we have no way of running it ourselves. Simplifying your example into a stand-alone snippet would come a long way to make it easier to pinpoint the source of the warnings.

**Comment by JobLeonard** [comment]
Ah, my apologies, I should have been more clear about why I closed this issue.

I'm pretty confident that I know the source of the warnings: the code is invoked within the Flask server, which I have set up to use threading.  So it actually is a correct warning - this wasn't what puzzled me. 

I opened this issue because it mysteriously *still* seemed to be running faster, but that turned out to be because of the early bail-out (presumably because of some interaction between how joblib was called within a server thread). So there are no bugs in joblib here as far as I can see, just in my own code.

Thanks for still having a look though, and sorry for the noise again!
